File: Creational/Prototype/DIY.java
package Creational.Prototype;

public class DIY {
    public static void main(String[] args) {
        var menu = new Menu();
        menu.copy(new Save());
    }
}

interface File{
    void add();
    File clone();
}

class Save implements File{
    private String fileName;

    public String getFileName() {
        return fileName;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }
    @Override
    public void add() {
        System.out.println("File saved!");
    }
    public File clone(){
        Save newLoad = new Save();
        newLoad.setFileName(fileName);
        return newLoad;
    }
    
}

class Menu{

    public void copy(File file){
        File newFile = file.clone();
        System.out.println("Original Save File: ");
        file.add();
        System.out.println("Copied Save File: ");
        newFile.add();

    }
}

File: TS/Creational/Prototype/DIY.ts
interface SaveFile {
  add(): void;
  clone(): Save;
}

class Save implements SaveFile {
  private fileName: string;

  public getFileName(): string {
    return this.fileName;
  }

  public setFileName(fileName: string): void {
    this.fileName = fileName;
  }

  public add(): void {
    console.log("File saved!");
  }

  public clone(): Save {
    const newLoad = new Save();
    newLoad.setFileName(this.fileName);
    return newLoad;
  }
}

class Menu {
  public copy(file: SaveFile): void {
    const newFile = file.clone();
    console.log("Original Save File: ");
    file.add();
    console.log("Copied Save File: ");
    newFile.add();
  }
}

// Main function
function DIYPrototype(): void {
  const menu = new Menu();
  menu.copy(new Save());
}

DIYPrototype();
